src/main/java/com/kaytat/simpleprotocolplayer/WorkerThreadPair.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2011 The Android Open Source Project
+ * Copyright (C) 2014 kaytat
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.kaytat.simpleprotocolplayer;
+
+import android.media.AudioFormat;
+import android.media.AudioManager;
+import android.media.AudioTrack;
+import android.os.Handler;
+import android.util.Log;
+import android.widget.Toast;
+
+/**
+ * 
+ * group everything belongs a stream together,makes multi stream easier
+ * including NetworkReadThread BufferToAudioTrackThread and AudioTrack
+ * 
+ */
+public class WorkerThreadPair {
+
+    static class ThreadStoppable extends Thread {
+        boolean running = true;
+
+        public void customStop() {
+            running = false;
+        }
+    }
+
+    private static final String TAG = WorkerThreadPair.class.getSimpleName();
+
+    private BufferToAudioTrackThread audioThread;
+    private NetworkReadThread networkThread;
+    AudioTrack mTrack;
+
+    public WorkerThreadPair(MusicService musicService, String serverAddr,
+            int serverPort, int sample_rate, boolean stereo, int buffer_ms) {
+
+        int format = stereo ? AudioFormat.CHANNEL_OUT_STEREO
+                : AudioFormat.CHANNEL_OUT_MONO;
+
+        // Sanitize input, just in case
+        if (sample_rate <= 0) {
+            sample_rate = MusicService.DEFAULT_SAMPLE_RATE;
+        }
+
+        if (buffer_ms <= 5) {
+            buffer_ms = MusicService.DEFAULT_BUFFER_MS;
+        }
+
+        int minBuf = AudioTrack.getMinBufferSize(sample_rate, format,
+                AudioFormat.ENCODING_PCM_16BIT);
+
+        packet_size = calcPacketSize(sample_rate, stereo, minBuf, buffer_ms);
+
+        for (int i = 0; i < NUM_PKTS; i++) {
+            byteArray[i] = new byte[packet_size];
+        }
+
+        // The agreement here is that mTrack will be shutdown by the helper
+        mTrack = new AudioTrack(AudioManager.STREAM_MUSIC, sample_rate, format,
+                AudioFormat.ENCODING_PCM_16BIT, minBuf, AudioTrack.MODE_STREAM);
+
+        audioThread = new BufferToAudioTrackThread(this, "audio_thread_of_"
+                + serverAddr + ":" + serverPort);
+        networkThread = new NetworkReadThread(this, serverAddr, serverPort,
+                "network_thread_of_" + serverAddr + ":" + serverPort);
```

Comment:
Although safe, there is an existing lint rule that complains if the tag is greater than 23 characters long.  And that would probably happen here.  To play 'nice' with the logger, can you try to reduce this?  Either limit the tag to be xyz.xyz.xyz.xyz:ppppppp (I think that's 23 chars) or use an index of some sort for the tag and dump the translation between index and ip/port at the beginning.